### ISSUE TITLE
addr2line can not parse relocatable object file

### ISSUE THREAD

**Issue by liutao-liu** [gimli-rs/addr2line#286]
The problem scenario is the same as that described in this issue：https://github.com/gimli-rs/gimli/issues/689

**Problem Description**
The behavior of addr2line is similar to that of example```simple-line.rs``` in gimli, and only the executable file can be parsed.
If the relocatable object file needs to be parsed and the file contains multiple units, the ```find_location_range``` function always returns the row of the first unit. For details about how to reproduce the problem, see the preceding issue.

**Modification Scheme**
The value is the same as ```dwarfdump.rs``` in Gimli. The function ```new_with_sup``` , need to relocate the debug.info section as follows:
![image](https://github.com/gimli-rs/addr2line/assets/10509166/b498f2b1-fb8d-49c4-a22f-d97bd8ebf0fe)

**Comment by philipc** [comment]
I don't think it makes sense to run addr2line on a relocatable object file. It's relocatable, so addresses haven't been assigned yet.

**Comment by liutao-liu** [comment]
> I don't think it makes sense to run addr2line on a relocatable object file. It's relocatable, so addresses haven't been assigned yet.
```/usr/bin/addr2line``` can work properly, similar to objdump:
![image](https://github.com/gimli-rs/addr2line/assets/10509166/d2ee5cb0-7feb-440d-88be-58e5fe0257c1)

**Comment by philipc** [comment]
And do you have an actual need for this? What is your use case?

**Comment by liutao-liu** [comment]
> And do you have an actual need for this? What is your use case?

The current application scenario is similar to that of the issue: gimli-rs/gimli#689.

In general, I need to parse the line number and starting address in the relocatable object.

**Comment by philipc** [comment]
> In general, I need to parse the line number and starting address in the relocatable object,similar to a.out generated above

Why do you need to do this? Where are you getting the address from that you are going to pass to addr2line? The file you are generating is no executable, so the address is not coming from executing this.

**Comment by liutao-liu** [comment]
> > In general, I need to parse the line number and starting address in the relocatable object,similar to a.out generated above
> 
> Why do you need to do this? Where are you getting the address from that you are going to pass to addr2line? The file you are generating is no executable, so the address is not coming from executing this.


Referring to ```dwarfdump.rs``` in gimli, reimplement ```read_address, read_offset, read_length, and read_sized_offset``` in ```gimli::Reader``` . The results returned by the four interfaces will be relocated.

![image](https://github.com/gimli-rs/addr2line/assets/10509166/1ec03318-f273-4b65-9636-0040a99661c9)

```/usr/bin/addr2line``` can work properly to parse the relocatable object, so I think this is normal and reasonable, so ```gimli-rs/addr2line```  should also be able to parse the relocatable object normally. After all, they both have the same name, addr2line.

**Comment by liutao-liu** [comment]
In addition, the example in the [issue](https://github.com/gimli-rs/gimli/issues/689) , seems unreasonable, because this example is my simplified application scenario, which facilitates communication with you about the problem scenario and problem reproduction.

**Comment by philipc** [comment]
I fully understand how the relocation code in gimli works, because I wrote it.

You're not answering my question. Why specifically do you need to be able to run addr2line for an address in a relocatable object? Where did that address come from? What led you to the point where you needed do to this?

To me, what you are doing is very strange. You are linking two files together to create a single relocatable object file, but calling the result `a.out`, which is name that is usually reserved for executable files. Are you sure that you know what you are doing? Please give an explanation.

**Comment by philipc** [comment]
> because this example is my simplified application scenario

I'm asking for more detail about your application scenario, because you simplified scenario doesn't make sense.

**Comment by liutao-liu** [comment]
> > because this example is my simplified application scenario
> 
> I'm asking for more detail about your application scenario, because your simplified scenario doesn't make sense.
The ```objdump```` command parses the relocatable object, so I run the address used in the ```/usr/bin/addr2line``` command from objdump.

![image](https://github.com/gimli-rs/addr2line/assets/10509166/a553066f-3132-4f75-ad64-29b1085dea3e)

**Comment by liutao-liu** [comment]
The ```objdump``` and ```addr2line``` commands are used only to prove that the relocatable object can be parsed to obtain ```line number``` and ```starting address```. So ```gimli/addr2line```, also called addr2line, should work properly.

**Comment by philipc** [comment]
Ok, but what are you trying to do? Why do you need to get an address from `objdump` and pass it to `addr2line`? Why do you need to use this library instead of `/usr/bin/addr2line`? Why do you need to use this library on a relocatable object?

This library is designed for symbolicating addresses that obtained from unwinding the stack during the execution of a program. For this use, handling relocations is never needed, so adding support for relocations is unneeded complexity. As such, I want to be convinced that this is a useful feature before adding this complexity.

Additionally, you could implement the relocation handling yourself by supplying your own `Reader` implementation via `Context::from_dwarf` instead of using `Context::new_with_sup`. This wouldn't need any changes to this library.

**Comment by liutao-liu** [comment]
The method you mentioned sounds feasible. I'll try it. Thank you very much for your quick reply.

**Comment by liutao-liu** [comment]
Done！Thank you again！
